OWHierarchicalClustering: Add labels to dendrogram

janezd opened the pull request:
##### Issue

Closes #1088.

##### Description of changes

<img width="812" alt="Screenshot 2019-06-06 at 12 17 26" src="https://user-images.githubusercontent.com/2387315/59026335-74d21200-8856-11e9-8ef4-3678e770fcd5.png">

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

janezd commented on Orange/widgets/unsupervised/owhierarchicalclustering.py:
```diff
@@ -462,29 +496,30 @@ def select_item(self, item, state):
             self.selectionChanged.emit()
 
+    @staticmethod
+    def _create_path(item, path):
```
This code is just factored out from lines 477 and 592. This change is not necessary; I've needed it in my initial attempt, but then kept it because it's better this way.

janezd commented on Orange/widgets/unsupervised/owhierarchicalclustering.py:
```diff
@@ -1766,7 +1795,8 @@ def __init__(self, parent=None, orientation=Qt.Vertical, value=0.0,
         self.setAcceptedMouseButtons(Qt.LeftButton)
-        self.setPen(make_pen(brush=QColor(50, 50, 50), width=1, cosmetic=False))
+        self.setPen(make_pen(brush=QColor(50, 50, 50), width=1, cosmetic=False,
+                             style=Qt.DashLine))
```
Cut-off line often crosses the new labels, which looked ugly. Dashed line looks better.

BlazZupan commented:
Nice, works well. When a cluster is a singlet, the distance of the text from the clustering line is disproportionally high. Consider C2 and C4 in the following clustering:

![image](https://user-images.githubusercontent.com/726604/59088498-f03fcc00-8907-11e9-99e5-9ce626e91f94.png)

Another minor change could be to position cluster labels below the line for the lower clustering branches. That is, in the example, you have provided above, C4 should be placed below the line so that the line does not cross the label. Using this rule, C1 in your visualization should be placed below the line as well. Whether or not to implement this is debatable, I like the consistency of labels being all above the clustering line while placing them above and below would reduce the clutter between the text and the graphics.

janezd commented:
I moved the labels for singlets.

For singlets I also moved the rectangle itself to the left to right-align it with other selection polygons. If you prefer what we had before, I can move them back.

Moving labels below the line, is not doable with this implementation of labels. I only know the shapes of polygons and not where in the structure they lie.
